### PR TITLE
[MRG+1] Made hashing module robust to dtype and string shared in memory

### PR DESCRIPTION
**Pull request by arthurmensch**
Shoulf fix bug #231, maybe #213 and nilearn/nilearn#771

### CODE REVIEW

**Comment by ogrisel** [comment]
Oops, I posted the previous comment with the wrong account...

**Comment by arthurmensch** [comment]
It was great hearing from circle ci himself! I am currently working on py3
compatibility

File: joblib/hashing.py [hunk 1]
```diff
@@ -77,6 +78,32 @@ def save(self, obj):
...
+
+        # The use of the Unpickler memo length as the memo key is just a
+        # convention.  The only requirement is that the memo values be unique.
+        # But there appears no advantage to any other scheme, and this
+        # scheme allows the Unpickler memo to be implemented as a plain (but
+        # growable) array, indexed by memo key.
+        if self.fast:
+            return
+        # Separate bytes and unicode shared by objects in the pickle graph,
+        # to avoid cache invalidation when unpickling a subobject of the graph
```
Comment (GaelVaroquaux) [inline joblib/hashing.py]: I think that you should state in the comment above that this is the only place where this method differs from the parent's method.

Comment (GaelVaroquaux) [inline joblib/hashing.py]: Even better, you could write this code at the beginning of the method, and call the parent's method after this code.

File: joblib/hashing.py [hunk 2]
```diff
@@ -146,7 +173,15 @@ def __init__(self, hash_name='md5', coerce_mmap=False):
         else:
             self._getbuffer = memoryview
 
-    def save(self, obj):
+    def persistent_id(self, obj):
+        # Separate dtypes shared by objects in the pickle graph,
+        # to avoid cache invalidation when unpickling a subobject of the graph
+        if isinstance(obj, self.np.dtype):
+            return obj.str
```
Comment (GaelVaroquaux) [inline joblib/hashing.py]: @ogrisel : returning a pickle.dumps here would be more safe, but slower. What do you think: should we do that, or check stick with the 'str', following a rule of thumb that "what reprs together hashes together"

File: joblib/hashing.py (hunk 1 above)
Comment (ogrisel) [inline joblib/hashing.py]: +1

File: joblib/hashing.py (hunk 2 above)
Comment (ogrisel) [inline joblib/hashing.py]:
```
%timeit dumps(a.dtype)
100000 loops, best of 3: 12.5 µs per loop
```

I don't think this is a problem.

However I don't understand the goal of this method. Where is it used?

File: joblib/test/test_hashing.py [hunk 3]
```diff
@@ -320,3 +322,18 @@ def test_set_hash():
...
+
+def test_string():
+    string = 'foo'
+    a = {string: 'bar'}
+    b = {string: 'bar'}
+    c = pickle.loads(pickle.dumps(b))
+    assert_equal(hash([a, b]), hash([a, c]))
+
+
+def test_dtype():
```
Comment (ogrisel) [inline joblib/test/test_hashing.py]: Tests that require numpy should be protected with `@with_numpy` to skip them when numpy is not installed:

```
from joblib.test.common import np, with_numpy
```

File: joblib/test/test_hashing.py [hunk 4]
```diff
@@ -320,3 +322,18 @@ def test_set_hash():
     b = k.f(a)
 
     nose.tools.assert_equal(hash(a), hash(b))
+
+
+def test_string():
```
Comment (GaelVaroquaux) [inline joblib/test/test_hashing.py]: Please explain, in a comment, what these tests actually test (the thing about the identity isn't trivial).

**Comment by GaelVaroquaux** [comment]
The PR looks good, but the comments that @ogrisel and myself made must be addressed before merging.

**Comment by arthurmensch** [comment]
Last Travis failure seems unrelated

**Comment by GaelVaroquaux** [comment]
The travis failure is not unrelated: it seems that you forgot to add a "with_numpy" to protect that test.

**Comment by GaelVaroquaux** [comment]
This looks good to me. +1 for merge

Great job, @arthurmensch !

File: joblib/hashing.py [hunk 5]
```diff
@@ -146,7 +152,15 @@ def __init__(self, hash_name='md5', coerce_mmap=False):
...
-    def save(self, obj):
+    def persistent_id(self, obj):
+        # Separate dtypes shared by objects in the pickle graph,
+        # to avoid cache invalidation when unpickling a subobject of the graph
+        if isinstance(obj, self.np.dtype):
+            return pickle.dumps(obj)
+        else:
+            return None
+
+    def save(self, obj, save_persistent_id=True):
```
Comment (ogrisel) [inline joblib/hashing.py]: The `save_persistent_id` kwarg is never used. Why is this introduced?

Comment (ogrisel) [inline joblib/hashing.py]: Ok I get it, it's required by python 3. I don't get it why it was not required before, but fair enough.

**Comment by ogrisel** [comment]
This looks good to me, merging.

**Comment by GaelVaroquaux** [comment]
Awesome work. This was super non trivial and it's really useful to have
it merged.

**Pull request merged**